docs/development/logging.md

```diff
@@ -0,0 +1,174 @@
+# Logging in Gardener Components
+
+This document aims at providing a general developer guideline on different aspects of logging practices and conventions used in the Gardener codebase.
+It contains mostly Gardener-specific points and references other existing and commonly accepted logging guidelines for general advice.
+Developers and reviewers should consult this guide when writing, refactoring and reviewing Gardener code.
+If parts are unclear or new learnings arise, this guide should be adapted accordingly.
+
+## Logging Libraries / Implementations
+
+Historically, Gardener components have been using [logrus](https://github.com/sirupsen/logrus).
+There is a global logrus logger ([`logger.Logger`](https://github.com/gardener/gardener/blob/626ba7c10e1150819b3905116d3988512c18c9ee/pkg/logger/logrus.go#L28)) that is initialized by components on startup and used across the codebase.
+In most places, it is used as a `printf`-style logger and only in some instances we make use of logrus' structured logging functionality.
+
+In the process of migrating our components to native controller-runtime components (see [gardener/gardener#4251](https://github.com/gardener/gardener/issues/4251)), we also want to make use of controller-runtime's built-in mechanisms for streamlined logging.
+controller-runtime uses [logr](https://github.com/go-logr/logr), a simple structured logging interface, for library-internal logging and logging in controllers.
+
+logr itself is only an interface and doesn't provide an implementation out of the box.
+Instead, it needs to be backed by a logging implementation like [zapr](https://github.com/go-logr/zapr). Code that uses the logr interface is thereby not tied to a specific logging implementation and makes the implementation easily exchangeable.
+controller-runtime already provides a [set of helpers](https://github.com/kubernetes-sigs/controller-runtime/tree/v0.11.0/pkg/log/zap) for constructing zapr loggers, i.e., logr loggers backed by [zap](https://github.com/uber-go/zap), which is a popular logging library in the go community.
+Hence, we are migrating our component logging from logrus to logr (backed by zap) as part of [gardener/gardener#4251](https://github.com/gardener/gardener/issues/4251).
+
+> ⚠️ `logger.Logger` (logrus logger) is deprecated in Gardener and shall not be used in new code – use logr loggers when writing new code!
```

Comment:
To really consider the logrus logger deprecated, I believe this guideline should contain some more information. I actually stumbled upon these things when switching #5150 to `logr.Logger`.
* How to actually use `logr.Logger` in gardenlet loggers? Some pointers to example code would be great.
* What to do with library packages that expect e.g. `logrus.FieldLogger` before they are refactored? E.g., construct a `logrus.FieldLogger` that "immitates" `logrus.Logger` just for these cases and put a TODO, etc.

Reply:
> * How to actually use `logr.Logger` in gardenlet loggers? Some pointers to example code would be great.

Hmm, I think this will be resolved once the first controllers are migrated. So I don't know if it's worth putting it into the guideline.

> * What to do with library packages that expect e.g. `logrus.FieldLogger` before they are refactored? E.g., construct a `logrus.FieldLogger` that "immitates" `logrus.Logger` just for these cases and put a TODO, etc.

This should also be resolved pretty soon in on of the upcoming PRs. Apart from that, I'm unsure what general advice to give for that case, because it will pretty much depend on the case. Also there shouldn't be many cases of this, I think.

Reply:
Ok, I added a short section on the migration itself including the pointers you mentioned above and the ones I sent you on Slack.
I think, this should be good enough for now :)